src/PureScript/Backend/Chez/Convert.purs

```diff
@@ -318,3 +322,17 @@ codegenPrimOp codegenEnv = case _ of
               S.List [ S.Identifier "scm:not", S.List [ opStringOrd o', x', y' ] ]
             _ ->
               S.List [ opStringOrd o', x', y' ]
+
+codegenCtorDef :: Ident -> Array String -> ChezExpr
+codegenCtorDef (Ident i) [] =
+  S.List
+    [ S.Identifier "scm:cons"
+    , S.List [ S.Identifier "scm:quote", S.Identifier i ]
+    , S.Identifier "scm:nil"
+    ]
+codegenCtorDef (Ident i) ss =
+  S.List
+    [ S.Identifier "scm:cons"
+    , S.List [ S.Identifier "scm:quote", S.Identifier i ]
+    , S.List $ [ S.Identifier "scm:vector" ] <> map S.Identifier ss
+    ]
```

Comment:
For constructors with no args, why not just use a symbol?

Per [Section 11.13](http://www.r6rs.org/final/r6rs.pdf):
> Vectors are heterogeneous structures whose elements are indexed by integers. A vector typically occupies **less space than a list of the same length**, and the **average time needed to access a randomly chosen element is typically less for the vector than for the list** (emphasis mine).

For constructors with args, why not use vectors here?

Reply:
> For constructors with no args, why not just use a symbol?

The following is based on assumptions that might be completely wrong as I'm not yet very familiar with purscript-backend-optimizer. Please feel free to correct me.

At the discrimination site between different constructors of the same type (OpIsTag), we're only provided with the constructor identifier. This information is not sufficient to choose whether to emit something like `(eq? tag 'Foo)` (for a constructor without args) or `(eq? (car x) 'Foo)` (for a constructor with arguments). Similar considerations apply for the constructor invocation (or lack thereof).

To do what you suggest we might need to keep the CtorMeta for all the types around. I'm not sure if that's a good tradeoff considering that in both `(define Nothing 'Nothing)` and `(define (lambda () 'Nothing))`, `'Nothing` is an interned symbol and no allocation is performed.

> For constructors with args, why not use vectors here?

So you're suggesting to emit `(vector 'Just value0)` instead of `('Just . (vector value0))`? That's a good point! I used that structure because it's quite common in Scheme to tag structures like that, by having a pair whose car is the tag. Maybe we might also want to discuss using records (cf. #84)?

Reply:
I think constructors may map to records better, since we also get predicate functions for free when checking for OpIsTag—that's my guess at least.

Reply:
@turlando One PR that was merged but not yet released is https://github.com/aristanetworks/purescript-backend-optimizer/pull/44. I was thinking we'd do something similar here. But looking at that PR again, it only works if all the constructors in a sum type have 0 args. If any of them have at least 1, then I'm not sure it will work.

Reply:
Thats correct, it only optimizes data types that are only enums. I would be shocked if there was a measurable advantage from a runtime/memory perspective since only one constructor is ever allocated, I think it’s mainly that it looks cleaner, is somewhat expected, and was not difficult to implement.

Reply:
> For constructors with no args, why not just use a symbol?

[bf70fee](https://github.com/purescm/purescm/pull/85/commits/bf70fee3ccb89f7dcfd5ce5726382397c8581e08) implements it.